node/src/components/consensus/era_supervisor.rs

```diff
@@ -463,6 +501,79 @@ where
     pub(crate) fn is_initialized(&self) -> bool {
         self.is_initialized
     }
+
+    fn handle_initialize_eras(
+        &mut self,
+        key_blocks: HashMap<EraId, BlockHeader>,
+        mut maybe_validators: Option<BTreeMap<PublicKey, U512>>,
```

Comment:
I really don't like this. Is it really worth sacrificing readability for saving a couple of bytes for the time of this method execution (which is like nanoseconds probably)?

Reply:
I wasn't thinking about performance at all: I rather thought it would be _more_ readable if the `Option` made explicit the distinction between the case where we do need and provide a validator set for the initial era, and where we don't.

But of course a validator set can never be empty, so using an empty one instead of `None` would work as well — so you'd prefer just having `BTreeMap<PublicKey, U512>` here?

Reply:
`Some(empty_tree)` is still wrong so we have two ways in which it can fail – `None` and `Some(empty)`. I would prefer if we had one.

Reply:
Yes… to my mind, the best way to express this would be if we had a `NonEmptyMap` type. Then it should be `Option<NonEmptyMap>`, because `None` expresses "no validator set" better than `Some(empty_map)`.

Anyway, I'll change it to `BTreeMap` and remove the `Option`. :+1:

Reply:
It also means we have to needlessly `clone` the map, though. :confused: 
Or `mem::replace` it with `BTreeMap::new()`.

Reply:
And I think we should have separate methods for the case when we need a validator map and when we don't need it :sweat_smile:

Reply:
> It also means we have to needlessly `clone` the map, though. 
> Or `mem::replace` it with `BTreeMap::new()`.

We can pass in `&BTreeMap` and let the client clone it if necessary, right?

Reply:
But the caller doesn't need the map anymore. It's the function itself that uses it in the loop, and it uses it in only one of the iterations, but Rust doesn't know that.

I'll give it some more thought. Maybe I can refactor it somehow…

Reply:
So you don't like the fact that it's going to be cloned … once?

Reply:
No, it's not a big deal. :+1:

Reply:
Done in https://github.com/CasperLabs/casper-node/pull/1023/commits/10ee3692c0bfc436cb953d2f6735070bd4f49f84.